graph mode: more docs for insert observers pass

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39828 [wip] qat graph mode: add combine_conv_bn pass
* #39837 eager qat: refactor conv-bn to make conv with scaled weights a func
* #39840 qat graph mode: make per channel observers scriptable
* #39796 qat graph mode: initial scaffolding for prepare_qat_script
* #39795 graph mode: add quantization type enum
* #39750 graph mode qat: make fake_quantize scriptable
* **#39739 graph mode: more docs for insert observers pass**
* #39683 graph mode: add docs to pre-calibration passes

Summary:

Adding more docs and examples to make code reasing easier for newcomers.

Test Plan:

CI, no logic changes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21975878](https://our.internmc.facebook.com/intern/diff/D21975878)

vkuzo committed:
graph mode: more docs for insert observers pass

Summary:

Adding more docs and examples to make code reasing easier for newcomers.

Test Plan:

CI, no logic changes

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_observers.cpp:
```diff
@@ -341,8 +345,12 @@ class InsertObserversHelper {
 
+  // Uses the state created by fillBoundaryValueMap and fillValueObserverMap
+  // to return an observer for a value, if it is needed.
   c10::optional<Module> getObserverFor(Value* v);
```
you can change this to `get observer module configured for the value`

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_observers.cpp:
```diff
@@ -305,6 +305,10 @@ class InsertObserversHelper {
+  // block_observed_values. If v is a part of a delayed observation pattern,
+  // record v's descendant (per delay rules) instead. The observers are inserted
+  // at a later stage by reading the state created by this function.
   void recordObserved(
```
value_to_observe stores the values that we'll need to insert observers for, block_observed_values are the values already observed in the current block

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_observers.cpp:
```diff
@@ -374,28 +382,75 @@ class InsertObserversHelper {
+  //   %x0: FooObserverModule,
+  //   %x1: BarObserverModule,
+  // }
   std::unordered_map<Value*, Module> observer_for_value_;
```
not sure if example is helpful for this one, I think you can say this is the map from a value to the observer module configured by user.

vkuzo replied:
agreed, it's probably not that helpful to people who work on this code full time.  It's targeted more to newcomers / people who need to sometimes touch this code.

jerryzh168 replied:
I mean the other examples look good, but this one is probably not as helpful as the others

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_observers.cpp:
```diff
@@ -374,28 +382,75 @@ class InsertObserversHelper {
+  //
+  // pass_through_value_map_ = {
+  //   %x1: [%x0],
+  //   %x2: [],
```
we don't have an entry for `%x2` I think

jerryzh168 reviewed (approved):
see some comments inline

vkuzo committed:
Update on "graph mode: more docs for insert observers pass"


Summary:

Adding more docs and examples to make code reasing easier for newcomers.

Test Plan:

CI, no logic changes

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "graph mode: more docs for insert observers pass"


Summary:

Adding more docs and examples to make code reasing easier for newcomers.

Test Plan:

CI, no logic changes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21975878](https://our.internmc.facebook.com/intern/diff/D21975878)

[ghstack-poisoned]

vkuzo committed:
Update on "graph mode: more docs for insert observers pass"


Summary:

Adding more docs and examples to make code reasing easier for newcomers.

Test Plan:

CI, no logic changes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21975878](https://our.internmc.facebook.com/intern/diff/D21975878)

[ghstack-poisoned]
